cvat/apps/quality_control/quality_reports.py

```diff
@@ -1000,6 +1004,21 @@ def _match_ann_type(self, t, *args):
         else:
             return None
 
+    def match_labels(self, item_a, item_b):
+        a_labels = self._get_ann_type(dm.AnnotationType.label, item_a)
+        b_labels = self._get_ann_type(dm.AnnotationType.label, item_b)
+
+        returned_values = _match_segments(
+            a_labels,
+            b_labels,
+            distance=lambda a, b: 1 - float(a is None or b is None),
+            label_matcher=lambda a, b: a.label == b.label,
+        )
+        if self.return_distances:
+            returned_values = returned_values + ({},)
```

Comment:
If you want to use `_match_segments`, it's probably better to use `self._match_segments` instead (and remove some code here). Otherwise the datumaro's one (which is inherited here) looks fine, it just needs some return value adjustment.

Reply:
Thanks for pointing out, I missed `self._match_segments` is the better version of the function, I will fix it. The values in the datumaro's one corresponding to frame results were incorrect.

Reply:
```py
a_unmatched = a_labels - b_labels
b_unmatched = b_labels - a_labels
```
This is from the `datumaro`'s `match_labels` implementation due to this line `a_unmatched` and `b_unmatched` are computed incorrectly and obviously due to this `mispred` isn't computed.

Reply:
Could you share an example when this is incorrect? "Mispred" is not defined for simple label comparison.

Reply:
Suppose there are two labels, "Dog" and "Cat". A frame is marked as "Dog" in GT Job and "Cat" in DS Job.
In this case,
[Expectation, if we use custom function]`matches`, `gt_unmatched` and `ds_unmatched` should be empty, whereas `mispred` contains information about that frame.
[Reality, if we use `datumaro`'s] `matches` and `mispred` are empty.
To make the results for the tag to make sense with existing parameters then, we have to make the implementation which can produce the outputs as in `Expectation`

Reply:
I can't totally agree about the expected result. It can be so if there is only 1 label per image, but there is no such restriction in CVAT or Datumaro.

Reply:
After you mentioned this, I tried it on several combinations of `gt_ann` and `ds_ann` for multi_label scenarios. It fails for the scenario, 
`gt_ann`: [tag_1, tag_2]
`ds_ann`: [tag_2, tag_3]
In such a case, it will try to compare tag_1 to tag_2 and tag_2 to tag_3, thus producing incorrect results.
Amazed by how you sensed all this just by looking at the code!

Reply:
Ahh, now, while trying to fix this. I got the feel of why mispred isn't there in datumaro's match_label.

`gt_ann`: [tag_1, tag_2, tag_3]
`ds_ann`: [tag_2, tag_4]
Now here, 
`matches`: [tag_2]
Comparing [tag_1, tag_3] and [tag_4] is ambiguous.

https://scikit-learn.org/stable/modules/generated/sklearn.metrics.multilabel_confusion_matrix.html Based on this also in the case of multi-label tag annotation, we have a class-wise confusion matrix, not a single confusion matrix. 

So, one thing we can do is to restrict this PR only to the annotations where each frame as single label. Or incorporate a label-wise (i.e., class-wise) confusion matrix in quality_report.

Reply:
> About the matrix - it's the same situation as with shapes, since there are no guarantees GT and DS have the same number of shapes or they overlap. It's already handled by the code. Per-label matrix is also there.

Thanks to this, it clicked that the distance function in the case of shapes is able to handle this. The distance function was not separate two tags of different labels, tags of the same labels are closer than of different labels.

> It's pretty simple if you look at this from this [perspective](https://en.wikipedia.org/wiki/Jaccard_index). Actually, it's used all over in this code - e.g. IoU, GT vs DS. Label matching as well.

Didn't got this...

Reply:
> Didn't got this...

This is to the comment:

> After you mentioned this, I tried it on several combinations of gt_ann and ds_ann for multi_label scenarios. It fails for the scenario,
gt_ann: [tag_1, tag_2]
ds_ann: [tag_2, tag_3]
In such a case, it will try to compare tag_1 to tag_2 and tag_2 to tag_3, thus producing incorrect results.
Amazed by how you sensed all this just by looking at the code!

For tags and other annotation types, we're just trying to find intersection of 2 sets of objects. Matches = intersection, gt_extra = gt - intersection, ds_extra = ds - intersection, just by definition. It's not related to the confusion matrix directly. Mispred is a part of intersection, that has different labels, it's separated to help with the following analysis.

Reply:
True, now it makes sense. The current code takes this into account.

Reply:
https://github.com/cvat-ai/cvat/pull/7582/commits/6ec3d2b2bf5bd5826eb57b25cc51a04215be37c6